src/renate/updaters/experimental/l2p.py

```diff
@@ -0,0 +1,289 @@
+from torch.nn import Parameter
+from torch.optim import Optimizer
+from torch.optim.lr_scheduler import _LRScheduler
+
+from renate import defaults
+from renate.benchmark.models.l2p import LearningToPromptTransformer
+from renate.models import RenateModule
+from renate.types import NestedTensors
+from renate.updaters.experimental.offline_er import OfflineExperienceReplayLearner
+from renate.updaters.learner import Learner
+from renate.updaters.model_updater import SingleTrainingLoopUpdater
+from renate.utils.misc import maybe_populate_mask_and_ignore_logits
+
+logger = logging.getLogger(__name__)
+
+
+class LearningToPromptLearner(Learner):
+    """Learner for learning to prompt
+
+    This is identical to the base learner with an addition of loss term.
+    TODO: Make this loss a component.
+
+    Args:
+        prompt_sim_loss_weight: Loss weight for the prompt key - image representation similarity
+    """
+
+    def __init__(
+        self,
+        prompt_sim_loss_weight: float = defaults.PROMPT_SIM_LOSS_WEIGHT,
+        **kwargs,
+    ) -> None:
+        assert isinstance(
+            kwargs["model"], LearningToPromptTransformer
+        ), f"{self.__class__.__name__} can only train a LearningToPromptTransformer model"
+        f"but got {type(kwargs['model'])}"
+        super().__init__(
+            **kwargs,
+        )
+        self.prompt_sim_loss_weight = prompt_sim_loss_weight
+        self._loss_collections["train_losses"].update({"key_sim_loss": torchmetrics.MeanMetric()})
+
+    def training_step(
+        self, batch: Tuple[NestedTensors, torch.Tensor], batch_idx: int
+    ) -> STEP_OUTPUT:
+        loss_dict = super().training_step(batch, batch_idx=batch_idx)
+        key_similarity = -1 * self.prompt_sim_loss_weight * self._model.similarity_score
+        loss_dict["loss"] += key_similarity
+        self._loss_collections["train_losses"]["key_sim_loss"](-key_similarity)
+        return loss_dict
+
+
+class LearningToPromptReplayLearner(OfflineExperienceReplayLearner):
+    """L2P with an off-line ER learner.
+
+    The model will be trained on weighted mixture of losses computed on the new data and a replay
+    buffer. In contrast to the online version, the buffer will only be updated after training has
+    terminated.
+
+    Args:
+        prompt_sim_loss_weight: Loss weight for the prompt key - image representation similarity
+    """
+
+    def __init__(
+        self,
+        prompt_sim_loss_weight: float = defaults.PROMPT_SIM_LOSS_WEIGHT,
+        **kwargs,
+    ) -> None:
+        assert isinstance(
+            kwargs["model"], LearningToPromptTransformer
+        ), f"{self.__class__.__name__}  can only train a LearningToPromptTransformer model"
+        f"but got {type(kwargs['model'])}"
+
+        super().__init__(**kwargs)
+        self.prompt_sim_loss_weight = prompt_sim_loss_weight
+        self._loss_collections["train_losses"].update({"key_sim_loss": torchmetrics.MeanMetric()})
+
+    def training_step(
```

Comment:
The comment on line 91 details why I do this reimplementation. 

TLDR: i need two forward calls. I can possibly hack together something by calling `super().forward` for each of them, but the hassle of managing the logging (because Offline-ER has its own) makes it not worth it.